### PR TITLE
Add task rollback if task download fails

### LINKED ISSUE #302: cache mechanism can become corrupt
Today, I had the following problem. When loading a task it kept on showing errors. Further inspection led me to the conclusion that this was due to the fact that the task arrived corrupt in the cache. 

We should try to fix this a little bit better, as this is hard to reproduce and due to the error message almost impossible to debug.

### PR DESCRIPTION
**Pull request by mfeurer**
This PR adds a rollback if task download fails (fixes #302).

### CODE REVIEW

File: openml/tasks/functions.py [hunk 1]
```diff
@@ -224,36 +225,90 @@ def get_task(task_id):
...
+def _remove_task_cache_dir(tid_cache_dir):
+    """Remove the task cache directory
+
+    This function is NOT thread/multiprocessing safe.
+
+    Parameters
+    ----------
+    """
+    try:
+        os.rmdir(tid_cache_dir)
```
Comment (amueller) [inline openml/tasks/functions.py]: why two different tries? Why not just shutil?

File: openml/tasks/functions.py [hunk 2]
```diff
@@ -224,36 +225,90 @@ def get_task(task_id):
...
+        Path of the created dataset cache directory.
+    """
+    task_cache_dir = os.path.join(
+        config.get_cache_directory(), "tasks", str(task_id)
+    )
+    try:
+        os.makedirs(task_cache_dir)
+    except (OSError, IOError):
+        # TODO add debug information!
+        pass
```
Comment (amueller) [inline openml/tasks/functions.py]: why should this pass? That seems like a severe error! (unless it already exists, but we're not testing that).

File: openml/tasks/functions.py [hunk 3]
```diff
@@ -224,36 +225,90 @@ def get_task(task_id):
...
+    tid_cache_dir = _create_task_cache_dir(task_id)
 
     with lockutils.external_lock(
             name='datasets.functions.get_dataset:%d' % task_id,
             lock_path=os.path.join(config.get_cache_directory(), 'locks'),
     ):
         try:
-            with io.open(xml_file, encoding='utf8') as fh:
-                task = _create_task_from_xml(fh.read())
+            task = _get_task_description(task_id)
```
Comment (amueller) [inline openml/tasks/functions.py]: I don't get why the logic changed so much here...

Comment (mfeurer) [reply openml/tasks/functions.py]: To make it look like getting a dataset. Things are now in their own little functions.

Comment (amueller) [reply openml/tasks/functions.py]: hm ok. I can try to go through the details. I mean it looks ok but I haven't looked through the details.

File: openml/tasks/functions.py (hunk 1 above)
Comment (mfeurer) [reply openml/tasks/functions.py]: Historic reasons... I changed this, and also changed this for the dataset functions.

File: openml/tasks/functions.py (hunk 2 above)
Comment (mfeurer) [reply openml/tasks/functions.py]: You're right. I just changed this, also for the dataset functions.

File: openml/tasks/functions.py (hunk 3 above)
Comment (mfeurer) [reply openml/tasks/functions.py]: The tests are passing and I just shuffled some lines, so don't think there's something weird going on here.

**Pull request merged**